### PR TITLE
[SPARK-33782][K8S][CORE]Place spark.files, spark.jars and spark.files under the current working directory on the driver in K8S cluster mode

### PR DESCRIPTION
**Pull request by pralabhkumar**
### What changes were proposed in this pull request?
This PR will place spark.files , spark.jars and spark.pyfiles to the current working directory on the driver in K8s cluster mode 

### Why are the changes needed?
This mimics the behaviour of Yarn and also helps user to access files from PWD . Also as mentioned in the jira 
By doing this, users can, for example, leverage PEX to manage Python dependences in Apache Spark:
```
pex pyspark==3.0.1 pyarrow==0.15.1 pandas==0.25.3 -o myarchive.pex
PYSPARK_PYTHON=./myarchive.pex spark-submit --files myarchive.pex
```

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Tested via unit test cases and also ran on local K8s cluster.

### CODE REVIEW

**Comment by pralabhkumar** [comment]
@HyukjinKwon  Please review , build is failing because of un related error (since its passing on local)

File: core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala [hunk 1]
```diff
@@ -486,6 +486,38 @@ class SparkSubmitSuite
     conf.get("spark.kubernetes.driver.container.image") should be ("bar")
   }
 
+  test("SPARK-33782 : handles k8s  files download to current directory") {
```
Comment (HyukjinKwon) [inline core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:489]:
```suggestion
  test("SPARK-33782 : handles k8s files download to current directory") {
```

File: core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala [hunk 2]
```diff
@@ -486,6 +486,38 @@ class SparkSubmitSuite
...
+      "/home/thejar.jar",
+      "arg1")
+    val appArgs = new SparkSubmitArguments(clArgs)
+    val (childArgs, classpath, conf, mainClass) = submit.prepareSubmitEnvironment(appArgs)
+    conf.get("spark.master") should be ("k8s://https://host:port")
+    conf.get("spark.executor.memory") should be ("5g")
+    conf.get("spark.driver.memory") should be ("4g")
+    conf.get("spark.kubernetes.namespace") should be ("spark")
+    conf.get("spark.kubernetes.driver.container.image") should be ("bar")
+    import java.nio.file.{Paths, Files}
```
Comment (HyukjinKwon) [inline core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:512]: Let's import this on the top

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 3]
```diff
@@ -381,45 +381,52 @@ private[spark] class SparkSubmit extends Logging {
...
+              val source = new File(localResources.getPath)
               val dest = new File(
                 ".",
                 if (resolvedUri.getFragment != null) resolvedUri.getFragment else source.getName)
               logInfo(
-                s"Unpacking an archive $resolvedUri " +
+                s"Files  $resolvedUri " +
                   s"from ${source.getAbsolutePath} to ${dest.getAbsolutePath}")
               Utils.deleteRecursively(dest)
               Utils.unpack(source, dest)
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:406]: Hm, why do we try this unpack for jars and files too? I think we should just call `downloadFileList` for them

**Comment by HyukjinKwon** [comment]
cc @dongjoon-hyun @tgravescs @Ngone51 FYI

**Comment by tgravescs** [comment]
> Does this PR introduce any user-facing change?
> No

Is this true?  If I specify a file to pass along and my code references it, what did it do before?   I know normally on yarn I would do some like ./XXXX. 

I'm a bit worried here about breaking people if we move location and they expected it somewhere else.

**Comment by pralabhkumar** [comment]
@tgravescs 

https://github.com/apache/spark/pull/30735/files#r540935585 , In case K8s files are not copied to current working directory . Files are available in local but not in current working directory for driver.

**Comment by pralabhkumar** [comment]
ping @dongjoon-hyun @Ngone51

**Comment by pralabhkumar** [comment]
ping @dongjoon-hyun @Ngone51

**Comment by pralabhkumar** [comment]
ping @dongjoon-hyun @Ngone51 . I think this is important feature to have the same behaviour as yarn . Please review.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 4]
```diff
@@ -381,45 +382,52 @@ private[spark] class SparkSubmit extends Logging {
...
                 ".",
                 if (resolvedUri.getFragment != null) resolvedUri.getFragment else source.getName)
               logInfo(
-                s"Unpacking an archive $resolvedUri " +
+                s"Files  $resolvedUri " +
                   s"from ${source.getAbsolutePath} to ${dest.getAbsolutePath}")
               Utils.deleteRecursively(dest)
-              Utils.unpack(source, dest)
-
+              if (isArchive) Utils.unpack(source, dest) else Files.copy(source.toPath, dest.toPath)
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:408]:
```suggestion
              if (isArchive) {
                Utils.unpack(source, dest)
              } else {
                Files.copy(source.toPath, dest.toPath)
              }
```

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 5]
```diff
@@ -381,45 +382,52 @@ private[spark] class SparkSubmit extends Logging {
...
+          downloadResourcesToCurrentDirectory(_, true)
+        }.orNull
+        val pyLocalFiles = Option(args.pyFiles).map {
+          downloadResourcesToCurrentDirectory(_)
         }.orNull
         args.files = filesLocalFiles
         args.archives = archiveLocalFiles
-        args.pyFiles = localPyFiles
+        args.pyFiles = pyLocalFiles
+        args.jars = jarsLocalJars
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:430]: `jarsLocalJars` -> `localJars`?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 6]
```diff
@@ -381,45 +382,52 @@ private[spark] class SparkSubmit extends Logging {
       localPyFiles = Option(args.pyFiles).map {
         downloadFileList(_, targetDir, sparkConf, hadoopConf)
       }.orNull
-
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:384]: revert unnecessary change?

File: core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala [hunk 7]
```diff
@@ -486,6 +486,41 @@ class SparkSubmitSuite
     conf.get("spark.kubernetes.driver.container.image") should be ("bar")
   }
 
+  test("SPARK-33782 : handles k8s files download to current directory") {
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:489]: nit:
```suggestion
  test("SPARK-33782: handles k8s files download to current directory") {
```

**Comment by Ngone51** [comment]
The change generally looks good to me. It'd be good if @dongjoon-hyun could take a look since he has more knowledge in K8s.

**Comment by pralabhkumar** [comment]
Thx @Ngone51  , for reviewing this , i'll do the suggested changes. 

@dongjoon-hyun Please review the PR .

**Comment by pralabhkumar** [comment]
Gentle ping @dongjoon-hyun

**Comment by pralabhkumar** [comment]
Gentle ping @dongjoon-hyun

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 5 above)
Comment (pralabhkumar) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:430]: Since localJars is already defined above , used this variable (which is simillar to filesLocalFiles)

**Comment by pralabhkumar** [comment]
@dongjoon-hyun , please review this , it will help to have K8s similar functionality as Yarn . Since this is target for Spark3.4 (as per jira) , IMHO , it would be great if u can spend some time on it . 

cc @HyukjinKwon (since jira is created by [HyukjinKwon])

**Comment by pralabhkumar** [comment]
@dongjoon-hyun , Have incorporated all the review comments , please look into the same.

**Comment by pralabhkumar** [comment]
@HyukjinKwon @Ngone51 @dongjoon-hyun 
Gentle ping to please review the same.

**Comment by pralabhkumar** [comment]
@HyukjinKwon  , can u please reviewed or get it reviewed. Since this is important jira (opened by you) and also fix parity with yarn .

**Comment by pralabhkumar** [comment]
@Yikun can u please review this PR .

**Review by holdenk — APPROVED** [review]
LGTM, if no one else has any concerns I'll try and merge it this week or next.

**Comment by pralabhkumar** [comment]
@HyukjinKwon  Thx for giving LGTM.  Please merge the PR

**Comment by holdenk** [comment]
Merged, thanks for the reminder.

**Comment by pralabhkumar** [comment]
Thx @holdenk for your help .

**Comment by holdenk** [comment]
Thanks for making the PR :D

**Comment by stijndehaes** [comment]
Maybe this is not the correct place to ask, but this PR had created some issues for our flow. Maybe our flow was not how things were intended to be used but it goes like this:
We submit jobs on kubernetes by baking in the jar in the docker image, we copy the jar in the docker image on the workdir location.
The result of this PR is that our jar is removed, because of the `Utils.deleteRecursively(dest)`.
Now a quick fix is to copy the jar somewhere else, but if this sounds like unintended behavior to you guys, I am willing to open up a PR to fix it.
I am however not sure what the intended behavior would be. Just remove this `Utils.deleteRecursively(dest)` sounds a bit too simple :)

**Comment by pm-nuance** [comment]
@pralabhkumar @HyukjinKwon @holdenk 
Facing issue with the new Spark version which is using Files.Copy in SparkSubmit.scala.
The latest change in the SparkSubmit.scala is causing the NoSuchFileException.
The mentioned "sample.jar" is present in the mentioned path /opt/spark/work-dir/sample.jar inside the docker image.
My same code is working fine with Spark 3.3.0.

**Here is the Error Message**
Files  local:///opt/spark/work-dir/sample.jar from /opt/spark/work-dir/sample.jar to /opt/spark/work-dir/./sample.jar
Exception in thread "main" java.nio.file.NoSuchFileException: /opt/spark/work-dir/sample.jar
        at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
        at sun.nio.fs.UnixCopyFile.copy(UnixCopyFile.java:526)
        at sun.nio.fs.UnixFileSystemProvider.copy(UnixFileSystemProvider.java:253)
        at java.nio.file.Files.copy(Files.java:1274)
        at org.apache.spark.deploy.SparkSubmit.$anonfun$prepareSubmitEnvironment$14(SparkSubmit.scala:437)
        at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:286)
        at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
        at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
        at scala.collection.TraversableLike.map(TraversableLike.scala:286)
        at scala.collection.TraversableLike.map$(TraversableLike.scala:279)
        at scala.collection.AbstractTraversable.map(Traversable.scala:108)
